tests/osqp_test.py

```diff
@@ -0,0 +1,511 @@
+from sklearn import svm
+
+from jaxopt import projection
+from jaxopt._src.base import KKTSolution
+from jaxopt._src.osqp import OSQP
+from jaxopt import QuadraticProgramming
+
+
+def get_random_osqp_problem(problem_size, eq_constraints, ineq_constraints):
+  assert problem_size >= eq_constraints  # very likely to be infeasible
+  onp.random.seed(problem_size + eq_constraints + ineq_constraints)
+  Q = onp.random.randn(problem_size, problem_size)
+  Q = Q.T.dot(Q)  # PSD matrix
+  c = onp.random.randn(problem_size)
+  A = onp.random.randn(ineq_constraints + eq_constraints, problem_size)
+  l = onp.random.randn(ineq_constraints)
+  u = l + jnp.abs(onp.random.randn(ineq_constraints))  # l < u
+  b = onp.random.randn(eq_constraints)  # Ax = b
+  l = jnp.concatenate([l, b])
+  u = jnp.concatenate([u, b])
+  return (Q, c), A, (l, u)
+
+
+def _from_osqp_form_to_default_form(Q, c, A, l, u):
+  """Return form used by QuadraticProgramming from OSQP form (support only matrices)."""
+  is_eq = l == u
+  is_ineq_l = jnp.logical_and(l != u, l != -jnp.inf)
+  is_ineq_u = jnp.logical_and(l != u, u != jnp.inf)
+  if jnp.any(is_eq):
+    A_eq, b = A[is_eq,:], l[is_eq]
+  else:
+    A_eq, b = jnp.zeros((1,len(c))), jnp.zeros((1,))
+  if jnp.any(is_ineq_l) and jnp.any(is_ineq_u):
+    G = jnp.concatenate([-A[is_ineq_l,:],A[is_ineq_u,:]])
+    h = jnp.concatenate([-l[is_ineq_l],u[is_ineq_u]])
+  elif jnp.any(is_ineq_l):
+    G, h = -A[is_ineq_l,:], -l[is_ineq_l]
+  elif jnp.any(is_ineq_u):
+    G, h = A[is_ineq_u,:], u[is_ineq_u]
+  else:
+    return (Q, c), (A_eq, b), None
+  return (Q, c), (A_eq, b), (G, h)
+
+
+class OSQPTest(jtu.JaxTestCase):
+
+  @parameterized.product(eq_ineq=[(0, 4), (4, 0), (4, 4)])
+  def test_small_qp(self, eq_ineq):
+    # Setup a random QP min_x 0.5*x'*Q*x + q'*x s.t. Ax = z; l <= z <= u;
+    eq_constraints, ineq_constraints = eq_ineq
+    onp.random.seed(42)
+    problem_size = 16
+    params_obj, params_eq, params_ineq = get_random_osqp_problem(problem_size, eq_constraints, ineq_constraints)
+    tol = 1e-5
+    osqp = OSQP(tol=tol)
+    params, state = osqp.run(None, params_obj, params_eq, params_ineq)
+    self.assertLessEqual(state.error, tol)
+    opt_error = osqp.l2_optimality_error(params, params_obj, params_eq, params_ineq)
+    self.assertAllClose(opt_error, 0.0, atol=1e-4)
+
+    def test_against_cvxpy(params_obj):
+      (Q, c), Ab, Gh = _from_osqp_form_to_default_form(params_obj[0], params_obj[1],
+                                                       params_eq, params_ineq[0], params_ineq[1])
+      Q = 0.5 * (Q + Q.T)
+      qp = QuadraticProgramming()
+      hyperparams = dict(params_obj=(Q, c), params_eq=Ab, params_ineq=Gh)
+      sol = qp.run(**hyperparams).params
+      return sol.primal
+
+    atol = 1e-4
+    cvx_primal = test_against_cvxpy(params_obj)
+    self.assertArraysAllClose(params.primal[0], cvx_primal, atol=atol)
+
+    def osqp_run(params_obj):
+      Q, c = params_obj
+      P = 0.5 * (Q + Q.T)
```

Comment:
Just noticed that this isn't used; is it supposed to be ?

Reply:
Yes indeed ! It should be `Q = 0.5 * (Q + Q.T)`. Thanks for the notification I will correct it in future PR.